### ISSUE TITLE
Linux support 

### ISSUE THREAD

**Issue by Rikitik** [BoyC/GW2TacO#8]
I offer $200 and ingame legendaries for somewhat native support. Wine is fine but it need a proper Transperancy implementation instead of compositor opcity work around

**Comment by coderedart** [comment]
in the reddit post of taco going open source, there was some discussion of someone wanting to attempt to integrate taco into gw2 like arcdps integrates with d3d9.dll. but as taco uses dx11, boyc said that the whole rendering layer will need to be rewritten with dx9 to integrate it like arcdps. but with the news of gw2 shifting to dx11, the option is back on table. although idk if it will work, just wanted to mention that.

if someone can get taco working on linux, that's great and you can give them money. but otherwise, i will release a barebones taco port for linux (first, it can just show markers and enable/disable them) within a week-ish. you can see the progress here, https://github.com/coderedart/jokolay .

**Comment by Technetium1** [comment]
I still haven't gotten this to work https://gitlab.com/namoninja/gw2taco_linux

**Comment by Rikitik** [comment]
Gw2 will possibly go to dx11(asuming they pull through) so arc will have to go to dx11 too if it does. Anyway this bounty will stay

**Comment by coderedart** [comment]
@Rikitik https://www.reddit.com/r/Guildwars2/comments/p3sioz/burrito_a_gw2_overlay_tool_for_linux/

Someone made it 🙂

**Comment by namoninja** [comment]
I've made a native WINE patch.

https://gitlab.com/namoninja/proton-nn

**Comment by icavalheiro** [comment]
now with wayland support transparency on KDE and Gnome, is this something we could revisit?

**Comment by BoyC** [comment]
How does that work? Is there a native linux gw2 client or is it running through some translation layer as well?

**Comment by namoninja** [comment]
@BoyC I ended up patching wine at source ([wine/wine-master](https://gitlab.winehq.org/wine/wine)) to implement `DwmExtendFrameIntoClientArea`, which ensures ARGB visuals, correct window styles and some additional input/z-order heuristics.

The project moved away from a **proton patch** ([Proton-NN](https://gitlab.com/namoninja/proton-nn)) to **wine patch** [wine-custom/dwm-overalys](https://gitlab.com/namoninja/wine-custom/-/tree/main/projects/dwm-overlays), however this does not look like something that will be approved by upstream or implemented anytime soon ([Merge Request #10180 - Desktop Window Manager (DWM) Overlay and Composition Support ](https://gitlab.winehq.org/wine/wine/-/merge_requests/10180)).

You can read about the full implementation details and build my patched version from my repo - [wine-custom/projects/dwm-overalys](https://gitlab.com/namoninja/wine-custom/-/tree/main/projects/dwm-overlays). It's a bit behind but the build tools anchor to the exact upstream tag that I worked from so it should build fine. I also planned to try to patch dxgi swapchains specifically to support taco but I ran out of time.

I started working on a proton patch from the upstream patch but I only wanted to focus on wayland support, which proton itself does not support well as during my tests there are downstream issues they have not patched with the window creation process. However I paused since there is not much interest in something like this. I started to look at providing a **Proton-GE** patch for user friendliness/ease of support but that is 2 levels down from wine upstream.

I would need to understand **Proton-GE**'s patches to see where they fixed proton's wayland issues and then reverse patch from my upstream one. I might try at some point but for now it's paused. The build tools are still there at the **Proton-NN** project and if I make any progress I will update there.

**Comment by BoyC** [comment]
@namoninja that doesn't answer my question. It's nice that you're doing work for getting transparency to work and you'll be able to create overlays that way but the main issue is still mumble link. I'm not sure if gw2 has a linux client, and if it does if it contains mumble link support. And if gw2 is run through wine/proton/whatever you'll need to figure out if the mumble shared memory file is working properly in order to get TacO to work. Transparency isn't enough.

**Comment by namoninja** [comment]
@BoyC mumble link has always worked with taco though, am I missing something? It was only a transparency issue.

There is no GW2 Linux client. Even the Mac client was decommissioned a long time ago. People are using Wine with tweaks or primarily Proton-GE to run GW2 on Linux.

GW2TacO issues were primarily isolated to 2 main issues. DirectComposition swapchains (newer Taco) and DWM Transparency. These features were/are stubbed in Wine and that's why TacO does not work (or somewhat partially works).

IIRC TacO is not able to start at all without MumbleLink and in my tests and other peoples runs TacO is able to start and establish the link, but fails at transparency and dxgi swapchain creation.

It's also important to ensure that GW2 is run under the same wine prefix as Taco otherwise the MumbleLink fails.

**Comment by namoninja** [comment]
Basically there is nothing Taco can change in its codebase to fix this issue apart from rewriting most of the core logic to be linux native (Vulkan/Wayland). Running GW2 or GW2TacO in whatever format depends on Wine, and both issues are there as stubs/lack of support.

There are other workarounds like DLL injections and Shaders, but if you want Linux support, then Wine needs to be patched. Hence my attempts.

If you are willing to write native linux calls via Vulkan, im not sure how much you will have to adapt in your app as Vulkan is the Linux equivalent of DirectX 12 so most of your calls/swapchain creations would need a linux counterpart. Secondly Wayland is difficult to work with due to its security protocols. Since TacO is a cross-process overlay im not sure you can create the right subsurface area to attach to GW2, at least my attempts failed via Wine's wayland driver.

And third, even if you manage to rewrite to linux code, GW2 will still be running via Wine so you would have to attach to that prefix somehow/wine's engine to ensure the link is attached (as you mentioned shared memory file).

**Comment by coderedart** [comment]
@icavalheiro What exactly changed in KDE/Gnome for this topic to be resurrected?

**Comment by icavalheiro** [comment]
Gnome 51 just launched yesterday and they have enabled proper transparency on wayland now, and KDE is also almost done with it as well.
But im not expert in that area, I just saw the gnome 51 patch notes and immediately remembered this thread.

**Comment by coderedart** [comment]
@icavalheiro  yeah, gnome 51 is about [background-blur effect](https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/5071). Transparency always worked on wayland/x11. That's how `Burrito`  works.

As the other comments said, the issue is `wine` not supporting creating a transparent window, and such a niche feature request would be really at the bottom of their priorities.

@namoninja For linux support, the easiest path is `glfw` for cross-platform windowing (supports transparency + mouse passthrough). `dx11` can be piped through `dxvk`. But porting a win32/dx11/visual-studio project like taco is basically a partial rewrite. It's plausible, but is still a lot of work for a tiny linux userbase. 

MumbleLink is a solved problem. Just like burritolink, you put a tiny arcdps-like dll in gw2 folder to be loaded by gw2 and every frame, that dll will just memcpy MumbleLink data to `z:/dev/shm/MumbleLink` which can be accessed by any linux app at `/dev/shm/MumbleLink`.
